# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/ipc/RemoteDecoderManagerChild.cpp
# Commit: 57607f8b6d8e
# Full Hash: 57607f8b6d8e1619b5c488494fd57547944c47b1
# Author: Dan Glastonbury <dan.glastonbury@gmail.com>
# Date: 2020-10-21 09:51:28
# Regressor Bug: 1595994
# File Overlap Count: 1
# Description:
#   Bug 1595994 - P7: Add Supports messages to PRemoteDecoderManager. r=kamidphish,nika,ipc-reviewers
#   
#   Add a synchronous Supports message to the IPDL PRemoteDecoderManager protocol so
#   decoder modules can query for playback support in the actual process that will
#   attempt to do the decoding.
# ==============================================================================

diff -r 1932f2f34993 -r 57607f8b6d8e dom/media/ipc/RemoteDecoderManagerChild.cpp
--- a/dom/media/ipc/RemoteDecoderManagerChild.cpp	Tue Oct 20 23:26:25 2020 +0000
+++ b/dom/media/ipc/RemoteDecoderManagerChild.cpp	Tue Oct 20 23:26:27 2020 +0000
@@ -63,6 +63,13 @@
 
 StaticRefPtr<ShutdownObserver> sObserver;
 
+static Maybe<layers::TextureFactoryIdentifier> MaybeTextureFactoryIdentifier(
+    const SupportDecoderParams& aParams) {
+  return aParams.mKnowsCompositor
+             ? Some(aParams.mKnowsCompositor->GetTextureFactoryIdentifier())
+             : Nothing();
+}
+
 /* static */
 void RemoteDecoderManagerChild::InitializeThread() {
   MOZ_ASSERT(NS_IsMainThread());
@@ -175,6 +182,40 @@
 }
 
 /* static */
+bool RemoteDecoderManagerChild::Supports(
+    RemoteDecodeIn aLocation, const SupportDecoderParams& aParams,
+    DecoderDoctorDiagnostics* aDiagnostics) {
+  bool supports = false;
+  DecoderDoctorDiagnostics diagnostics;
+
+  nsCOMPtr<nsISerialEventTarget> managerThread = GetManagerThread();
+  if (managerThread) {
+    RefPtr<Runnable> task =
+        NS_NewRunnableFunction("RemoteDecoderManager::Supports", [&]() {
+          auto* rdm = GetSingleton(aLocation);
+          const auto& trackInfo = aParams.mConfig;
+          if (trackInfo.GetAsVideoInfo()) {
+            VideoDecoderInfoIPDL info(*trackInfo.GetAsVideoInfo(),
+                                      aParams.mRate.mValue);
+            Unused << rdm->SendSupports(info,
+                                        MaybeTextureFactoryIdentifier(aParams),
+                                        &supports, &diagnostics);
+          } else if (trackInfo.GetAsAudioInfo()) {
+            Unused << rdm->SendSupports(*trackInfo.GetAsAudioInfo(), Nothing(),
+                                        &supports, &diagnostics);
+          }
+        });
+    SyncRunnable::DispatchToThread(managerThread, task);
+  }
+
+  if (aDiagnostics) {
+    *aDiagnostics = diagnostics;
+  }
+
+  return supports;
+}
+
+/* static */
 already_AddRefed<MediaDataDecoder>
 RemoteDecoderManagerChild::CreateAudioDecoder(
     const CreateDecoderParams& aParams) {